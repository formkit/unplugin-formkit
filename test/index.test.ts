import { fileURLToPath } from 'node:url'
import { describe, expect, it } from 'vitest'
import unplugin from '../src/vite'
import { readFileSync } from 'fs'

type Transformer = {
  load: (id: string) => Promise<string>
  transform: (code: string, id: string) => Promise<{ code: string; map?: any }>
}
const aboutSFCFile = readFileSync('./playground/src/pages/about.vue', 'utf-8')
const contactSFCFile = readFileSync(
  './playground/src/pages/contact.vue',
  'utf-8',
)
const rootDir = fileURLToPath(new URL('..', import.meta.url))
const plugin: Transformer = unplugin() as Transformer

describe('virtual configuration file', () => {
  const loadConfig = (...args: Parameters<typeof unplugin>) =>
    Promise.resolve(
      args.length
        ? (unplugin(...args) as Transformer).load('virtual:formkit-config')
        : plugin.load('virtual:formkit-config'),
    ).then((r) => r.replace(rootDir, '~/'))

  it('generates correct virtual configuration with default options', async () => {
    expect(await loadConfig()).toMatchSnapshot()
  })

  it('generates correct configuration when default config is disabled', async () => {
    expect(await loadConfig({ defaultConfig: false })).toMatchSnapshot()
  })

  it('generates correct configuration when custom config is available', async () => {
    expect(
      await loadConfig({
        configFile: './playground/formkit.config',
      }),
    ).toMatchSnapshot()
  })

  it('generates correct configuration when custom config is available and default config is disabled', async () => {
    expect(
      await loadConfig({
        configFile: './playground/formkit.config',
        defaultConfig: false,
      }),
    ).toMatchSnapshot()
  })
})

describe('vue file transformations', () => {
  it('injects the template block into an normally structured sfc', async () => {
    expect(
      (
        await plugin.transform(
          `<template>
  <FormKit />
</template>`,
          'test.vue',
        )
      ).code,
    ).toMatchSnapshot()
  })

  it('injects inside root node if there is one', async () => {
    expect(
      (
        await plugin.transform(
          `<template>
    <div class="fizzbuzz">
      <FormKit />
    </div>
  </template>`,
          'test.vue',
        )
      ).code,
    ).toMatchSnapshot()
  })

  it('injects inside root node with full sfc', async () => {
    expect(
      (
        await plugin.transform(
          `<script lang="ts" setup>
function handleLoginSubmit(values: any) {
  window.alert("You are logged in. Credentials: \n" + JSON.stringify(values));
}
</script>

<template>
  <div>
    <FormKit type="form" submit-label="login" @submit="handleLoginSubmit">
      <FormKit type="email" label="Email" name="email" />
      <FormKit type="password" label="Password" name="password" />
    </FormKit>
  </div>
</template>
`,
          'test.vue',
        )
      ).code,
    ).toMatchSnapshot()
  })

  it('injects inside root node with multiple child elements', async () => {
    expect(
      (
        await plugin.transform(
          `<script lang="ts" setup>
function handleLoginSubmit(values: any) {
  window.alert("You are logged in. Credentials: \n" + JSON.stringify(values));
}
</script>

<template>
  <div>
    <main>
    <p>
    <FormKit type="form" submit-label="login" @submit="handleLoginSubmit">
      <FormKit type="email" label="Email" name="email" />
      <FormKit type="password" label="Password" name="password" />
    </FormKit>
    </p>
    </main>
    <div class="filler">Here we go</div>
  </div>
</template>
`,
          'test.vue',
        )
      ).code,
    ).toMatchSnapshot()
  })

  it('injects import into script setup block', async () => {
    expect(
      (await plugin.transform(aboutSFCFile, 'about.vue')).code,
    ).toMatchSnapshot()
  })

  it('injects setup block when using options api', async () => {
    expect(
      (await plugin.transform(contactSFCFile, 'about.vue')).code,
    ).toMatchSnapshot()
  })
})
