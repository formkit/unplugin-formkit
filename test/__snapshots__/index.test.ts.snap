// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`virtual configuration file > generates correct configuration when custom config is available 1`] = `
"import { defaultConfig } from \\"@formkit/vue\\";
import _config from \\"~/playground/formkit.config.ts\\";
const config = typeof _config === 'function' ? _config() : _config;
export default defaultConfig(config);"
`;

exports[`virtual configuration file > generates correct configuration when custom config is available and default config is disabled 1`] = `
"import _config from \\"~/playground/formkit.config.ts\\";
const config = typeof _config === 'function' ? _config() : _config;
export default config;"
`;

exports[`virtual configuration file > generates correct configuration when default config is disabled 1`] = `
"const config = {};
export default config;"
`;

exports[`virtual configuration file > generates correct virtual configuration with default options 1`] = `
"import { defaultConfig } from \\"@formkit/vue\\";
const config = {};
export default defaultConfig(config);"
`;

exports[`vue file transformations > injects import into script setup block 1`] = `
"<script setup lang=\\"ts\\">
import { FormKitProvider } from \\"@formkit/vue\\";
import __formkitConfig from \\"virtual:formkit-config\\";
import { FormKit } from '@formkit/vue'
</script>

<template>
  
<FormKitProvider :config=\\"__formkitConfig\\">
<FormKit
    type=\\"text\\"
    label=\\"Your name\\"
    help=\\"Enter your name\\"
  />
</FormKitProvider>

</template>"
`;

exports[`vue file transformations > injects inside root node if there is one 1`] = `
"<script setup>
import { FormKitProvider } from \\"@formkit/vue\\";
import __formkitConfig from \\"virtual:formkit-config\\";
</script>
<template>
    <div class=\\"fizzbuzz\\">
      
<FormKitProvider :config=\\"__formkitConfig\\">
<FormKit />
</FormKitProvider>

    </div>
  </template>"
`;

exports[`vue file transformations > injects inside root node with full sfc 1`] = `
"<script lang=\\"ts\\" setup>
import { FormKitProvider } from \\"@formkit/vue\\";
import __formkitConfig from \\"virtual:formkit-config\\";
function handleLoginSubmit(values: any) {
  window.alert(\\"You are logged in. Credentials: 
\\" + JSON.stringify(values));
}
</script>

<template>
  <div>
    
<FormKitProvider :config=\\"__formkitConfig\\">
<FormKit type=\\"form\\" submit-label=\\"login\\" @submit=\\"handleLoginSubmit\\">
      <FormKit type=\\"email\\" label=\\"Email\\" name=\\"email\\" />
      <FormKit type=\\"password\\" label=\\"Password\\" name=\\"password\\" />
    </FormKit>
</FormKitProvider>

  </div>
</template>
"
`;

exports[`vue file transformations > injects inside root node with multiple child elements 1`] = `
"<script lang=\\"ts\\" setup>
import { FormKitProvider } from \\"@formkit/vue\\";
import __formkitConfig from \\"virtual:formkit-config\\";
function handleLoginSubmit(values: any) {
  window.alert(\\"You are logged in. Credentials: 
\\" + JSON.stringify(values));
}
</script>

<template>
  <div>
    
<FormKitProvider :config=\\"__formkitConfig\\">
<main>
    <p>
    <FormKit type=\\"form\\" submit-label=\\"login\\" @submit=\\"handleLoginSubmit\\">
      <FormKit type=\\"email\\" label=\\"Email\\" name=\\"email\\" />
      <FormKit type=\\"password\\" label=\\"Password\\" name=\\"password\\" />
    </FormKit>
    </p>
    </main>
    <div class=\\"filler\\">Here we go</div>
</FormKitProvider>

  </div>
</template>
"
`;

exports[`vue file transformations > injects setup block when using options api 1`] = `
"<script setup lang=\\"ts\\">
import { FormKitProvider } from \\"@formkit/vue\\";
import __formkitConfig from \\"virtual:formkit-config\\";
</script>
<script lang=\\"ts\\">
import { FormKit } from '@formkit/vue'

export default {
  components: {
    FormKit
  }
}
</script>

<template>
  <div>
    
<FormKitProvider :config=\\"__formkitConfig\\">
<h1>Nothing to see here</h1>
    <FormKit type=\\"text\\" label=\\"Check me out\\" />
</FormKitProvider>

  </div>
</template>"
`;

exports[`vue file transformations > injects the template block into an normally structured sfc 1`] = `
"<script setup>
import { FormKitProvider } from \\"@formkit/vue\\";
import __formkitConfig from \\"virtual:formkit-config\\";
</script>
<template>
  
<FormKitProvider :config=\\"__formkitConfig\\">
<FormKit />
</FormKitProvider>

</template>"
`;
